src/reflect/scala/reflect/runtime/JavaMirrors.scala

```diff
@@ -962,8 +962,9 @@ private[scala] trait JavaMirrors extends internal.SymbolTable with api.JavaUnive
       val enclosingClass = jclazz.getEnclosingClass
       var prefix = if (enclosingClass != null) enclosingClass.getName else ""
       val isObject = owner.isModuleClass && !owner.isPackageClass
-      if (isObject && !prefix.endsWith(nme.MODULE_SUFFIX_STRING)) prefix += nme.MODULE_SUFFIX_STRING
-      assert(jclazz.getName.startsWith(prefix))
+      val ownerIsJavaClass = owner.isJava && !owner.isPackageClass
+      if (isObject && !prefix.endsWith(nme.MODULE_SUFFIX_STRING) || ownerIsJavaClass) prefix += nme.MODULE_SUFFIX_STRING
+      assert(jclazz.getName.startsWith(prefix), (jclazz.getName, prefix))
```

Comment:
I'm quite confused about the intent here. It seems it's still buggy with Scala-defined classes:

```
➜  sandbox git:(be39803acc) ✗ cat A.java
package p;

public class A {
  public class B { };
  public static class C { };
}
➜  sandbox git:(be39803acc) ✗ javac -d . A.java
➜  sandbox git:(be39803acc) ✗ scala-pr 5978


import scala.reflect.runtime.universe._
val m = scala.runtime.ScalaRunTime.ensureAccessible(rootMirror.getClass.getDeclaredMethod("scalaSimpleName", classOf[Class[_]]))
def n[C: reflect.ClassTag] = m.invoke(rootMirror, implicitly[reflect.ClassTag[C]].runtimeClass).asInstanceOf[TypeName]


scala> :paste -raw
// Entering paste mode (ctrl-D to finish)

package u;
class A { class B }
object T { class B }

// Exiting paste mode, now interpreting.

scala> n[u.A#B]
res1: reflect.runtime.universe.TypeName = $B

scala> n[u.T.B]
res2: reflect.runtime.universe.TypeName = B

scala> n[p.A]
res3: reflect.runtime.universe.TypeName = A

scala> n[p.A#B]
res4: reflect.runtime.universe.TypeName = B

scala> n[p.A.C]
res5: reflect.runtime.universe.TypeName = C
```